pkg/proxy/iptables/proxier.go

```diff
@@ -1393,16 +1414,26 @@ func (proxier *Proxier) syncProxyRules() {
 			// Balancing rules in the per-service chain.
 			args = append(args[:0], "-A", string(svcChain))
 			args = proxier.appendServiceCommentLocked(args, svcNameString)
-			if i < (n - 1) {
+			if i < (numReadyEndpoints - 1) {
 				// Each rule is a probabilistic match.
 				args = append(args,
 					"-m", "statistic",
 					"--mode", "random",
-					"--probability", proxier.probability(n-i))
+					"--probability", proxier.probability(numReadyEndpoints-i))
 			}
 			// The final (or only if n == 1) rule is a guaranteed match.
 			args = append(args, "-j", string(endpointChain))
 			writeLine(proxier.natRules, args...)
+		}
+
+		// Every endpoint gets a chain, regardless of its state. This is required later since we may
+		// want to jump to endpoint chains that are terminating.
+		for i, endpointChain := range endpointChains {
```

Comment:
Does this mean we're adding significantly more chains (`O(endpoints)`)?  Or do we have per endpoint chains elsewhere?  What's the performance impact of this?

Reply:
Or can i safely assume this is NOT adding more per endpoint chains?  The comment is confusing.  Are we going from `O(ready endpoints)` chains to `O(endpoints)` chains?

Reply:
I think this actually reduces total number of iptable rules. There are actually two sets of rules that exist today:
1. The entire endpoint chains containing DNAT rules to a pod
2. The service chain which randomly picks one of the endpoint chains from 1. 

This change preserves the full endpoint chain from 1, but it updates 2. to only include ready endpoints or terminating endpoints. 

There is a performance penalty here for kube-proxy's sync loop though, because it now does two loops per Service, once for `endpointChains` and once for `readyEndpointChains`, but this seems negligible to me.

Reply:
Previously we only made chains for ready endpoints, and not-ready endpoints were just skipped.  Now we have chains for all endpoints.  If you assume that most endpoints are ready most of the time, this seems OK.  I don't see how this is less rules, though?

This updates the XLB chain, not the main service chain, and the whole point of this is that it can include terminating endpoints when it used to include nothing - so strictly MORE rules.  Right?

Reply:
> This updates the XLB chain, not the main service chain, and the whole point of this is that it can include terminating endpoints when it used to include nothing - so strictly MORE rules. Right?

At some point this PR updated the main service chain cause we wanted to fallback for `internalTrafficPolicy` as well, but I think we're now saying we only want it for `externalTrafficPolicy`, so yes more chains in that case.

Reply:
To close this out - we will make more chains after this change but it should not be significant because most chains are ready most of the time, anyway?

Reply:
I think ultimately we are going to be adding more chains, but the # of rules evaluated should not change since only ALL ready or ALL terminating endpoints are evaluated based on the probalistic load balancing in the service chain.